composer/callbacks/early_stopper.py

```diff
@@ -0,0 +1,130 @@
+            of the monitored metric, otherwise defaults to numpy.greater
+        min_delta (float, optional): An optional float that requires a new value to exceed the best value by at
+            least that amount. Defaults to 0.
+        patience (int | str | Time, optional): The interval of time the monitored metric can not improve without stopping
+            training. Defaults to 1 epoch.
+        
+    """
+
+    def __init__(
+        self,
+        monitor: str,
+        dataloader_label: str,
+        comp: Callable = None,
+        min_delta: float = 0.0,
+        patience: Union[int, str, Time] = 1,
+    ):
+        self.monitor = monitor
+        self.dataloader_label = dataloader_label
+        self.comp = comp
+        self.min_delta = abs(min_delta)
+        if self.comp is None:
+            if any(substr in monitor.lower() for substr in ["loss", "error", "perplexity"]):
+                self.comp = np.less
+            else:
+                self.comp = np.greater
+
+        self.best = None
+        self.best_occurred = None
+
+        if isinstance(patience, str):
+            self.patience = Time.from_timestring(patience)
+        elif isinstance(patience, int):
+            self.patience = Time(patience, TimeUnit.EPOCH)
+        else:
+            self.patience = patience
+            if self.patience.unit not in (TimeUnit.EPOCH, TimeUnit.BATCH):
+                raise ValueError("If `patience` is an instance of Time, it must have units of EPOCH or BATCH.")
+
+    def _get_monitored_metric(self, state: State):
+        if self.dataloader_label in state.current_metrics:
+            if self.monitor in state.current_metrics[self.dataloader_label]:
+                return state.current_metrics[self.dataloader_label][self.monitor]
+        raise ValueError(f"Couldn't find the metric {self.monitor} with the dataloader label {self.dataloader_label}."
+                         "Check that the dataloader_label is set to 'eval', 'train' or the evaluator name.")
+
+    def _update_stopper_state(self, state: State):
+        metric_val = self._get_monitored_metric(state)
+
+        if self.best is None:
+            self.best = metric_val
+            self.best_occurred = state.timer.get_timestamp()
+        elif self.comp(metric_val, self.best) and abs(metric_val - self.best) > self.min_delta:
+            self.best = metric_val
+            self.best_occurred = state.timer.get_timestamp()
+
+        if self.patience.unit == TimeUnit.EPOCH:
+            if int(state.timer.epoch) - int(self.best_occurred.epoch) > int(self.patience):
+                state.max_duration = state.timer.batch
+        elif self.patience.unit == TimeUnit.BATCH:
+            if int(state.timer.batch_in_epoch) - int(self.best_occurred.batch_in_epoch) > int(self.patience):
+                state.max_duration = state.timer.batch
+        else:
+            raise ValueError(f"The units of `patience` should be EPOCH or BATCH.")
```

Comment:
i _think_ `Timer` supports subtraction  and comparisons that would simplify this code (but @ravi-mosaicml is the expert here).

Also, you don't want to use `batch_in_epoch` -- what if patience is `100ba`, but that can span an epoch boundary?